# ZBIW-Zertifikatskurs "Data Librarian"
# Modul 2.1: Strukturierte Daten und Metadaten – Pandas Tutorial
Pandas ist ein flexibles und einfach zu bedienendes Open-Source Werkzeug zur Datenanalyse und -manipulation, das auf der Programmiersprache Python aufbaut.
Es vereinfacht das Arbeiten mit tabellarischen Daten, vor allem durch Methoden zum Import und Export einer Reihe von Dateiformaten, Übersichtlichkeit und integrierte Funktionalitäten zur Datenvisualisierung.

## Import von Pandas

In [1]:
# Pandas wird meist unter dem alias "pd" importiert
import pandas as pd

In [2]:
# Wenn pandas nicht installiert ist, kann es mit pip bzw. conda installiert werden
!pip install pandas

# Abhängig vom verwendeten Betriebssystem muss pip3 spezifiziert oder eine andere Formulierung genutz werden. 
# Mehr Informationen zur Installation unter https://pandas.pydata.org/docs/getting_started/install.html. 

## Dateiformate in Pandas – Series und DataFrame Objekte
Pandas verfügt über zwei grundlegende Objekttypen. Das erste ist das Series Objekt, das im Grunde wie eine Liste (Eselsbrücke: "Eine Serie von Einträgen") mit Indices funktioniert. Ein DataFrame aggregiert eine beliebige Anzahl von Series Objekten zu einer tabellarischen Struktur. Die Series Objekte stellen dabei die Spalten der Tabelle da.

### Series Objekte

In [3]:
list_of_strings = ["Value 1", "Value 2", "Value 3", "Value 4"]
list_of_integers = [1, 3, 6, 7]

In [4]:
series_of_strings = pd.Series(list_of_strings)

In [5]:
print(series_of_strings)

0    Value 1
1    Value 2
2    Value 3
3    Value 4
dtype: object


Ein Series Objekt kann, im Gegensatz zu einer einfachen Liste, auch labels haben:

In [6]:
labeled_series = pd.Series(list_of_integers, index = list_of_strings)

In [7]:
print(labeled_series)

Value 1    1
Value 2    3
Value 3    6
Value 4    7
dtype: int64


Entsprechend kann auch über ein dictionary ein Series Objekt erzeugt werden:

In [8]:
verkaeufe = {"Tag 1": 120, "Tag 2": 330, "Tag 3": 250, "Tag 4": 270}

series_verkäufe = pd.Series(verkaeufe)

print(series_verkäufe)

Tag 1    120
Tag 2    330
Tag 3    250
Tag 4    270
dtype: int64


In [9]:
# Zugriff auf Values über index Name
series_verkäufe["Tag 2"]

330

In [10]:
# Zugriff auf Values über index Position
series_verkäufe[1]

330

In [11]:
# Unterschied zu Dictionary: Ordered List
verkaeufe[1]

KeyError: 1

Auf die Werte eines Series Objektes kann auf verschiedene Weisen zugegriffen werden:

In [12]:
# die "to_list()" Funktion liefert die Elemente eines Series Objektes als Liste:
series_verkäufe.to_list()

[120, 330, 250, 270]

In [13]:
# Series Objekte sind iterierbar
for value in series_verkäufe:
    print(value)

120
330
250
270


### DataFrame Objekte 

#### Erstellung aus Listen

In [14]:
data1 = {
  "list1": [420, 380, 390],
  "list2": [50, 40, 45]
}

df_1 = pd.DataFrame(data1)


df_1

list1  list2
0    420     50
1    380     40
2    390     45

Listen müssen die selbe Länge haben! (Arrays sind eine besondere Form von Listen, die das NumPy Modul nutzt, das von Pandas genutzt wird)

In [15]:
data2 = {
  "list1": [420, 380, 390],
  "list2": [50, 40, 45],
  "list3": [1, 2, 3, 4, 5 , 6, 7]
}

df_2 = pd.DataFrame(data2)


df_2

ValueError: All arrays must be of the same length

#### Erstellung aus Dictionaries

Listen müssen von selber länge sein, Einzelwerte werden für alle Einträge des Dictionaries eingetragen.

In [16]:
data_dict = {
 "date" : "31.01.2022",
 "time" : ["08:22", "08:55", "09:21", "10:40", "11:30"],
 "value" : [12, 23, 3, 12, 3]    
}

df_3 = pd.DataFrame(data_dict)

In [17]:
df_3

date   time  value
0  31.01.2022  08:22     12
1  31.01.2022  08:55     23
2  31.01.2022  09:21      3
3  31.01.2022  10:40     12
4  31.01.2022  11:30      3

DataFrames können auch aus mehreren Dictionaries erstellt werden. Dann ist es auch möglich, Dictionaries unterschiedlicher Längen zusammenzuführen.

In [18]:
merlin = {"Name": "Merlin", "Tierart": "Katze", "Alter" : 3, "Farbe": "schwarz"}
otto = {"Name": "Otto", "Tierart": "Hund", "Alter" : 5, "Farbe": "braun"}
renate = {"Name": "Renate", "Tierart": "Fisch", "Alter" : 2, "Farbe": "rot"}

In [19]:
df_tiere = pd.DataFrame([merlin, otto, renate])

In [20]:
df_tiere

Name Tierart  Alter    Farbe
0  Merlin   Katze      3  schwarz
1    Otto    Hund      5    braun
2  Renate   Fisch      2      rot

In [21]:
verkaeufe = {"Tag 1": 120, "Tag 2": 330, "Tag 3": 250, "Tag 4": 270}
einnahmen = {"Tag 1": 450, "Tag 2": 1210, "Tag 3": 760, "Tag 4": 950, "Tag 5": 350}

# erstellen aus Liste von dictionaries:
df_4 = pd.DataFrame([verkaeufe, einnahmen], index = ["verkäufe", "einnahmen"])

In [22]:
df_4

Tag 1  Tag 2  Tag 3  Tag 4  Tag 5
verkäufe     120    330    250    270    NaN
einnahmen    450   1210    760    950  350.0

Das DataFrame wird mit "NaN", not a number aufgefüllt. Dieser wert ist ein float, also eine Kommazahl.

In [23]:
# mit ".T" kann ein Dataframe transponiert werden, wir tauschen also Zeilen und Spalten
df_4_transposed = df_4.T

In [24]:
df_4_transposed

verkäufe  einnahmen
Tag 1     120.0      450.0
Tag 2     330.0     1210.0
Tag 3     250.0      760.0
Tag 4     270.0      950.0
Tag 5       NaN      350.0

Über .info() und .describe() lassen sich einfach Informationen über DataFrames abrufen:

In [25]:
df_4_transposed.info()

<class 'pandas.core.frame.DataFrame'>
Index: 5 entries, Tag 1 to Tag 5
Data columns (total 2 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   verkäufe   4 non-null      float64
 1   einnahmen  5 non-null      float64
dtypes: float64(2)
memory usage: 292.0+ bytes


In [26]:
df_4_transposed.describe()

verkäufe    einnahmen
count    4.00000     5.000000
mean   242.50000   744.000000
std     88.45903   354.090384
min    120.00000   350.000000
25%    217.50000   450.000000
50%    260.00000   760.000000
75%    285.00000   950.000000
max    330.00000  1210.000000

### Wie DataFrames und Series Objekte zusammenhängen
Wie schon erwähnt, sind Dataframes "zusammengefügte" Series Objekte. Auf diese kann über den Namen der Spalte zugegriffen werden.

In [27]:
df_4_transposed.verkäufe

Tag 1    120.0
Tag 2    330.0
Tag 3    250.0
Tag 4    270.0
Tag 5      NaN
Name: verkäufe, dtype: float64

In [28]:
df_4_transposed["verkäufe"]

Tag 1    120.0
Tag 2    330.0
Tag 3    250.0
Tag 4    270.0
Tag 5      NaN
Name: verkäufe, dtype: float64

Die Spalten sind Series Objekte:

In [29]:
type(df_4_transposed.verkäufe)

pandas.core.series.Series

In [30]:
type(df_4_transposed["verkäufe"])

pandas.core.series.Series

#### Zugriff auf Spalten- und Zeilennamen (Indices) eines Dataframes

In [31]:
# Zugriff auf die Namen aller Spalten:
df_4_transposed.columns

Index(['verkäufe', 'einnahmen'], dtype='object')

In [32]:
# auch hier lässt sich "to_list()" verwenden: 
df_4_transposed.columns.to_list()

['verkäufe', 'einnahmen']

In [33]:
# Ähnlich lassen sich die Indices abrufen:
df_4_transposed.index

Index(['Tag 1', 'Tag 2', 'Tag 3', 'Tag 4', 'Tag 5'], dtype='object')

In [34]:
df_4_transposed.index.to_list()

['Tag 1', 'Tag 2', 'Tag 3', 'Tag 4', 'Tag 5']

## Einlesen von Dateien mit "read"
Pandas stellt eine Reihe von Methoden zur Verfügung, mit denen ohne Probleme eine Vielzahl verschiedener Datentypen eingelesen werden können. 

### CSV einlesen

In [35]:
df = pd.read_csv("data/nyt_bestsellers.csv")

In [36]:
df[:10]

1        I Love Dad with The Very Hungry Caterpillar  children
0   2                   The Wonderful Things You Will Be  children
1   3     Dr. Seuss's I Love Pop!: A Celebration of Dads  children
2   4                                 Dragons Love Tacos  children
3   5                           How to Babysit a Grandpa  children
4   6                                    I Wish You More  children
5   7                                      Grumpy Monkey  children
6   8                           The Day the Crayons Quit  children
7   9                                         Dear Girl,  children
8  10  Rosie Revere, Engineer (Questioneers Collectio...  children
9  11              Brawl of the Wild (Dog Man Series #6)  children

#### Spaltennamen und Indexspalte
In der [Dokumentation](https://pandas.pydata.org/docs/reference/api/pandas.read_csv.html) ist nachlesbar, welche Optionen es beim Einlesen gibt.

In [37]:
# names und index_col nutzen
df = pd.read_csv("data/nyt_bestsellers.csv", names = ["index_number", "title", "category"], index_col=0)
df

title  category
index_number                                                             
1                   I Love Dad with The Very Hungry Caterpillar  children
2                              The Wonderful Things You Will Be  children
3                Dr. Seuss's I Love Pop!: A Celebration of Dads  children
4                                            Dragons Love Tacos  children
5                                      How to Babysit a Grandpa  children
...                                                         ...       ...
96                      Everything Is F*cked: A Book about Hope      misc
97            Girl, Stop Apologizing: A Shame-Free Plan for ...      misc
98            You Are a Badass: How to Stop Doubting Your Gr...      misc
99            Dare to Lead: Brave Work. Tough Conversations....      misc
100           Medical Medium Celery Juice: The Most Powerful...      misc

[100 rows x 2 columns]

In [38]:
!head -1 data/search_Bibliothek.csv

BibliographyType,ISBN,Identifier,Author,Title,Journal,Volume,Number,Month,Pages,Year,Address,Note,URL,Booktitle,Chapter,Edition,Series,Editor,Publisher,ReportType,Howpublished,Institution,Organizations,School,Annote,Custom1,Custom2,Custom3,Custom4,Custom5


Vorsicht mit Separatoren: Pandas ist oft, vor allem wenn Kommata als Trennzeichen genutzt werden, nicht in der Lage, CSV Dateien korrekt einzulesen:

In [39]:
path = 'data/search_Bibliothek.csv'
df = pd.read_csv(path)

ParserError: Error tokenizing data. C error: Expected 31 fields in line 21, saw 34


In Zeile 21 sind im Text Kommata enthalten, was zu einer falschen Anzal erkannter Spalten führt.

In [40]:
!head -20 data/search_Bibliothek.csv | tail +20

7,"","ernst_unwahrscheinlichkeit_2018","Ernst, Wolfgang","Die Unwahrscheinlichkeit von Wissenstradition und die Beharrlichkeit der Bibliothek","Bibliothek Forschung und Praxis",42,2,"","379--386",2018,"","","https://www.degruyter.com/view/j/bfup.2018.42.issue-2/bfp-2018-0038/bfp-2018-0038.xml","","","","","","","","","","","","","Mit dem Wandel der Leitmedien von der Buchform zur Zeitform ändert sich der Auftrag der Bibliothek oder lässt diesen zumindest deutlicher erscheinen: Sie hat ihre Rolle im kybernetischen Denken der Wissenszirkulation zu aktualisieren als Zeitkanal zwischen materieller Entropie und ordnungsbewahrender Negentropie. Zwar ist der Hypertext eine aktuelle Alternative zum raumbezogenen Gedächtnis und eröffnet neue Optionen der Wissensnavigation, doch angesichts der Zerstreuung in Netzarchitekturen bedarf es der Bibliothek, um dem Verlust der Nachhaltigkeit von Online-Wissen katechontisch (verzögernd) entgegenzuwirken.","","bibliotheken","",""


In [41]:
# Kommas als Seperatoren schwierig wenn Zellen Kommas enthalten
!head -20 data/search_Bibliothek.csv | tail +20 | tr -cd ',' | wc -c

      33


Es gibt zwar Möglichkeiten, das Problem zu umgehen (z.B. über die Verwendung von "quotechar"), jedoch bietet es sich an, auf hierarchische Dateiformate wie JSON zurückzugreifen, wenn diese Verfügbar sind.

### JSON einlesen

In [47]:
# Einlesen einer JSON datei:
df_nyt = pd.read_json("data/nyt_bestseller.json")

In [48]:
df_nyt

_id  \
0   {'$oid': '5b4aa4ead3089013507db18b'}   
1   {'$oid': '5b4aa4ead3089013507db18c'}   
2   {'$oid': '5b4aa4ead3089013507db18d'}   
3   {'$oid': '5b4aa4ead3089013507db18e'}   
4   {'$oid': '5b4aa4ead3089013507db18f'}   
..                                   ...   
95  {'$oid': '5b4aa4ead3089013507db1ea'}   
96  {'$oid': '5b4aa4ead3089013507db1eb'}   
97  {'$oid': '5b4aa4ead3089013507db1ec'}   
98  {'$oid': '5b4aa4ead3089013507db1ed'}   
99  {'$oid': '5b4aa4ead3089013507db1ee'}   

                               bestsellers_date  \
0   {'$date': {'$numberLong': '1211587200000'}}   
1   {'$date': {'$numberLong': '1211587200000'}}   
2   {'$date': {'$numberLong': '1211587200000'}}   
3   {'$date': {'$numberLong': '1211587200000'}}   
4   {'$date': {'$numberLong': '1211587200000'}}   
..                                          ...   
95  {'$date': {'$numberLong': '1214006400000'}}   
96  {'$date': {'$numberLong': '1214006400000'}}   
97  {'$date': {'$numberLong': '1214006400000'}}   
98  {'$date': {'$numberLong': '1214006400000'}}   
99  {'$date': {'$numberLong': '1214006400000'}}   

                                 published_date  \
0   {'$date': {'$numberLong': '1212883200000'}}   
1   {'$date': {'$numberLong': '1212883200000'}}   
2   {'$date': {'$numberLong': '1212883200000'}}   
3   {'$date': {'$numberLong': '1212883200000'}}   
4   {'$date': {'$numberLong': '1212883200000'}}   
..                                          ...   
95  {'$date': {'$numberLong': '1215302400000'}}   
96  {'$date': {'$numberLong': '1215302400000'}}   
97  {'$date': {'$numberLong': '1215302400000'}}   
98  {'$date': {'$numberLong': '1215302400000'}}   
99  {'$date': {'$numberLong': '1215302400000'}}   

                                   amazon_product_url  \
0   http://www.amazon.com/Odd-Hours-Dean-Koontz/dp...   
1   http://www.amazon.com/The-Host-Novel-Stephenie...   
2   http://www.amazon.com/Love-Youre-With-Emily-Gi...   
3   http://www.amazon.com/The-Front-Garano-Patrici...   
4   http://www.amazon.com/Snuff-Chuck-Palahniuk/dp...   
..                                                ...   
95  http://www.amazon.com/Blood-Anita-Blake-Vampir...   
96  http://www.amazon.com/Snuff-Chuck-Palahniuk/dp...   
97  http://www.amazon.com/Death-Honor-Bound-Book/d...   
98  http://www.amazon.com/The-Spies-Warsaw-A-Novel...   
99  http://www.amazon.com/Phantom-Prey-John-Sandfo...   

                                        author  \
0                                Dean R Koontz   
1                              Stephenie Meyer   
2                                 Emily Giffin   
3                            Patricia Cornwell   
4                              Chuck Palahniuk   
..                                         ...   
95                          Laurell K Hamilton   
96                             Chuck Palahniuk   
97  W E B Griffin and William E Butterworth IV   
98                                  Alan Furst   
99                               John Sandford   

                                          description  \
0   Odd Thomas, who can communicate with the dead,...   
1   Aliens have taken control of the minds and bod...   
2   A woman's happy marriage is shaken when she en...   
3   A Massachusetts state investigator and his tea...   
4   An aging porn queens aims to cap her career by...   
..                                                ...   
95  The vampire hunter Anita Blake is involved in ...   
96  An aging porn queens aims to cap her career by...   
97  A Marine pilot spies in  Argentina in 1943; th...   
98       Intrigue in Poland just before World War II.   
99  The Minneapolis detective Lucas Davenport inve...   

                         price      publisher                     title  \
0         {'$numberInt': '27'}         Bantam                 ODD HOURS   
1   {'$numberDouble': '25.99'}  Little, Brown                  THE HOST   
2   {'$numberDouble': '24.95'}   St. Martin's  LOVE THE ONE YOU'RE WITH   
3   {'$numb

In [49]:
df_nyt.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100 entries, 0 to 99
Data columns (total 12 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   _id                 100 non-null    object
 1   bestsellers_date    100 non-null    object
 2   published_date      100 non-null    object
 3   amazon_product_url  100 non-null    object
 4   author              100 non-null    object
 5   description         100 non-null    object
 6   price               100 non-null    object
 7   publisher           100 non-null    object
 8   title               100 non-null    object
 9   rank                100 non-null    object
 10  rank_last_week      100 non-null    object
 11  weeks_on_list       100 non-null    object
dtypes: object(12)
memory usage: 9.5+ KB


Wir können noch keine Aussage über Datentypen treffen.

In [50]:
df_nyt.bestsellers_date[3]

{'$date': {'$numberLong': '1211587200000'}}

In [51]:
type(df_nyt.bestsellers_date[3])

dict

#### Bibliografische Informationen von Bibsonomy 
Als nächstes betrachten wir einen typischen Fall für den Import von bibliografischen Informationen. Wir lesen das Ergebnis einer Literatursuche bei Bibsonomy über das requests Modul ein.

In [52]:
# einlesen von Daten aus Bibsonomy
import requests
url = "https://www.bibsonomy.org/json/search/Bibliothek?items=1000"
result = requests.get(url) # result ist ein requests.models.Response Objekt
data = result.json() # das Response-Objekt bietet diese nützliche Methode an

Wie zuvor haben wir hier ein Dictionary, das drei Listen von Dictionaries enthält. Wir interessieren uns für den Wert des Keys "items", der die Suchergebnisse enthält:

In [53]:
data.keys()

dict_keys(['types', 'properties', 'items'])

In [54]:
type(data["items"])

list

In [55]:
# Hier zum Beispiel der 6. Eintrag: 
data["items"][5]

{'type': 'Bookmark',
 'id': 'https://www.bibsonomy.org/url/57a6c84d63a662d6fdda4b1b1155d68f/schneeschmelze',
 'tags': ['bibliothek'],
 'intraHash': '57a6c84d63a662d6fdda4b1b1155d68f',
 'label': 'Oodi-Bibliothek in Helsinki: Bollwerk gegen Populismus - taz.de',
 'user': 'schneeschmelze',
 'description': 'Der Common Ground in der modernen demokratischen Stadtgesellschaft: Die Finnen bauen die Bibliothek der Zukunft, digital und analog zugleich.',
 'date': '2018-12-17 21:06:43',
 'changeDate': '2018-12-17 21:06:43',
 'count': 1,
 'url': 'https://www.taz.de/Oodi-Bibliothek-in-Helsinki/!5556260/'}

Da es sich hier um eine Liste von Dictionaries handelt, können wir diese ganz einfach direkt mit Pandas als DataFrame einlesen.

In [58]:
df_bibsonomy = pd.DataFrame(data["items"])

In [59]:
df_bibsonomy

type                                                 id  \
0        Bookmark  https://www.bibsonomy.org/url/8dccb0fb43c1baa0...   
1        Bookmark  https://www.bibsonomy.org/url/6e4d27522857a23d...   
2        Bookmark  https://www.bibsonomy.org/url/74a579dd9b58dd57...   
3        Bookmark  https://www.bibsonomy.org/url/561cc0370c975e9d...   
4        Bookmark  https://www.bibsonomy.org/url/d062f38f9cf0a22e...   
...           ...                                                ...   
1995  Publication  https://www.bibsonomy.org/bibtex/278b2b1e22f9e...   
1996  Publication  https://www.bibsonomy.org/bibtex/241a87e3c1b43...   
1997  Publication  https://www.bibsonomy.org/bibtex/27e036a030b41...   
1998  Publication  https://www.bibsonomy.org/bibtex/211ecaafeb093...   
1999  Publication  https://www.bibsonomy.org/bibtex/29fcf69db60dd...   

                                                   tags  \
0                              [bibliothek, treffpunkt]   
1                                          [Bibliothek]   
2         [bibliothek, library, lang:de, referendariat]   
3     [bibliothek, wirtschaftswissenschaften, inform...   
4                                  [Bibliothek, BibNZZ]   
...                                                 ...   
1995  [Multinationales_Unternehmen, Theorie_der_Wirt...   
1996                                         [imported]   
1997                                         [imported]   
1998                                         [imported]   
1999  [filmtheorie, eisenstein, vertov, film, kino, ...   

                             intraHash  \
0     8dccb0fb43c1baa049c04533af10c807   
1     6e4d27522857a23d27d0fc781601e288   
2     74a579dd9b58dd57e61cbc58f2b354f2   
3     561cc0370c975e9d9143e81e9d901138   
4     d062f38f9cf0a22e25bffb32fb25e4ad   
...                                ...   
1995  78b2b1e22f9e497948dca06c78aee2a5   
1996  41a87e3c1b43161496cee15a9f6d8c45   
1997  7e036a030b4138d76543885c2bca4b53   
1998  11ecaafeb0931cb85e9c43c97fa46806   
1999  9fcf69db60dd700317bf3a1b9893fd2d   

                                                  label          user  \
0                     Bibliothek: Treffpunkt Bibliothek         hotho   
1                                            Bibliothek   stadlmairma   
2                                            Bibliothek       lambert   
3                             Bibliothek - Fit4Research         wdees   
4                                 Die dritte Bibliothek        szylla   
...                                                 ...           ...   
1995  Staat, multinationale Konzerne und politische ...  fbw_hannover   
1996  Demokratische Wirtschaftspolitik gegen Marktma...  fbw_hannover   
1997                        Für den Frieden produzieren  fbw_hannover   
1998                         Lateinische Umgangssprache  filologanoga   
1999                        Texte zur Theorie des Films         antje   

                                       description                 date  \
0     Treffpunkt Bibliothek 24. - 31. Oktober 2010  2010-07-18 21:30:25   
1                                                   2020-09-22 15:08:13   
2                                                   2005-06-22 15:33:30   
3                                                   2016-05-09 12:23:32   
4                 Bibliothek von Mircea Cartarescu  2007-12-03 10:43:42   
...                                            ...                  ...   
1995                                                2009-08-21 09:46:50   
1996                                                2009-08-21 09:46:50   
1997                                                2009-08-21 09:46:50   
1998                                                2011-08-05 10:55:22   
1999                                                2011-05-10 09:22:37   

               changeDate  count  ... imtm-inventarnummer eventtitle venue  \
0     2010-07-18 21:30:25      1  ...                 NaN        NaN   NaN   


### Andere Dateiformate

Beispielsweise mit read_excel() und read_xml() lassen sich noch andere Dateiformate als DataFrame lesen. Eine vollständige Übersicht findet sich [hier](https://pandas.pydata.org/docs/reference/io.html).

## Daten mit Pandas Auswählen und Bereinigen
Im folgenden betrachten wir einige Möglichkeiten, Daten Auszuwählen und zu Filtern.

### Spalten auswählen

In [63]:
# Wir betrachten die Spaltennamen:
df_bibsonomy.columns

Index(['type', 'id', 'tags', 'intraHash', 'label', 'user', 'description',
       'date', 'changeDate', 'count', 'url', 'interHash', 'pub-type', 'year',
       'author', 'authors', 'note', 'language', 'bibtexKey', 'publisher',
       'address', 'shorttitle', 'lccn', 'journal', 'volume', 'number', 'pages',
       'abstract', 'booktitle', 'series', 'editor', 'editors', 'doi', 'issn',
       'isbn', 'refid', 'file', 'urldate', 'eissn', 'subtitle', 'pagetotal',
       'ppn_gvk', 'location', 'ee', 'library', 'citeulike-article-id',
       'priority', 'comment', 'copyright', 'dnbtitleid', 'posted-at', 'tech',
       'ty', 'date-added', 'date-modified', 'title1', 'totalpages', 'ISBN',
       'pubs', 'hyphenation', 'booksubtitle', 'vgwortseiten', 'owner',
       'timestamp', 'username', 'intrahash', 'interhash', 'groups', 'price',
       'collaborator', 'size', 'holdings', 'collection', 'bdsk-url-1',
       'partauthors', 'uniqueid', 'shortsubtitle', 'format', 'titlestatement',
       'citeulik

In [62]:
# Die Spalte "author" enthält anscheinend in vielen Zeilen keine Angabe:
df_bibsonomy.author

0                      NaN
1                      NaN
2                      NaN
3                      NaN
4                      NaN
               ...        
1995    [Heinz-Jürgen Axt]
1996                  [ ?]
1997                   NaN
1998       [J. B. Hofmann]
1999                   NaN
Name: author, Length: 2000, dtype: object

In [64]:
# Ebenso die Spalte "year":
df_bibsonomy.year

0        NaN
1        NaN
2        NaN
3        NaN
4        NaN
        ... 
1995    1978
1996    1981
1997    1981
1998    1926
1999    1998
Name: year, Length: 2000, dtype: object

In [67]:
# Die Spalte Label enthält die Titel der Veröffentlichungen:
df_bibsonomy.label

0                       Bibliothek: Treffpunkt Bibliothek
1                                              Bibliothek
2                                              Bibliothek
3                               Bibliothek - Fit4Research
4                                   Die dritte Bibliothek
                              ...                        
1995    Staat, multinationale Konzerne und politische ...
1996    Demokratische Wirtschaftspolitik gegen Marktma...
1997                          Für den Frieden produzieren
1998                           Lateinische Umgangssprache
1999                          Texte zur Theorie des Films
Name: label, Length: 2000, dtype: object

Über das Format  
```
DataFrame[["v1", "v2", "v3", "v4"]]
```
Lässt sich eine Ansicht, also eine art Kopie des DataFrames ausgeben, das nur die angegebenen Spalten enthält.

In [69]:
df_bibsonomy[["label", "author", "year", "type"]]

label              author  \
0                     Bibliothek: Treffpunkt Bibliothek                 NaN   
1                                            Bibliothek                 NaN   
2                                            Bibliothek                 NaN   
3                             Bibliothek - Fit4Research                 NaN   
4                                 Die dritte Bibliothek                 NaN   
...                                                 ...                 ...   
1995  Staat, multinationale Konzerne und politische ...  [Heinz-Jürgen Axt]   
1996  Demokratische Wirtschaftspolitik gegen Marktma...                [ ?]   
1997                        Für den Frieden produzieren                 NaN   
1998                         Lateinische Umgangssprache     [J. B. Hofmann]   
1999                        Texte zur Theorie des Films                 NaN   

      year         type  
0      NaN     Bookmark  
1      NaN     Bookmark  
2      NaN     Bookmark  
3      NaN     Bookmark  
4      NaN     Bookmark  
...    ...          ...  
1995  1978  Publication  
1996  1981  Publication  
1997  1981  Publication  
1998  1926  Publication  
1999  1998  Publication  

[2000 rows x 4 columns]

Dabei wird aber NICHT das ursprüngliche DataFrame überschrieben:

In [70]:
df_bibsonomy

type                                                 id  \
0        Bookmark  https://www.bibsonomy.org/url/8dccb0fb43c1baa0...   
1        Bookmark  https://www.bibsonomy.org/url/6e4d27522857a23d...   
2        Bookmark  https://www.bibsonomy.org/url/74a579dd9b58dd57...   
3        Bookmark  https://www.bibsonomy.org/url/561cc0370c975e9d...   
4        Bookmark  https://www.bibsonomy.org/url/d062f38f9cf0a22e...   
...           ...                                                ...   
1995  Publication  https://www.bibsonomy.org/bibtex/278b2b1e22f9e...   
1996  Publication  https://www.bibsonomy.org/bibtex/241a87e3c1b43...   
1997  Publication  https://www.bibsonomy.org/bibtex/27e036a030b41...   
1998  Publication  https://www.bibsonomy.org/bibtex/211ecaafeb093...   
1999  Publication  https://www.bibsonomy.org/bibtex/29fcf69db60dd...   

                                                   tags  \
0                              [bibliothek, treffpunkt]   
1                                          [Bibliothek]   
2         [bibliothek, library, lang:de, referendariat]   
3     [bibliothek, wirtschaftswissenschaften, inform...   
4                                  [Bibliothek, BibNZZ]   
...                                                 ...   
1995  [Multinationales_Unternehmen, Theorie_der_Wirt...   
1996                                         [imported]   
1997                                         [imported]   
1998                                         [imported]   
1999  [filmtheorie, eisenstein, vertov, film, kino, ...   

                             intraHash  \
0     8dccb0fb43c1baa049c04533af10c807   
1     6e4d27522857a23d27d0fc781601e288   
2     74a579dd9b58dd57e61cbc58f2b354f2   
3     561cc0370c975e9d9143e81e9d901138   
4     d062f38f9cf0a22e25bffb32fb25e4ad   
...                                ...   
1995  78b2b1e22f9e497948dca06c78aee2a5   
1996  41a87e3c1b43161496cee15a9f6d8c45   
1997  7e036a030b4138d76543885c2bca4b53   
1998  11ecaafeb0931cb85e9c43c97fa46806   
1999  9fcf69db60dd700317bf3a1b9893fd2d   

                                                  label          user  \
0                     Bibliothek: Treffpunkt Bibliothek         hotho   
1                                            Bibliothek   stadlmairma   
2                                            Bibliothek       lambert   
3                             Bibliothek - Fit4Research         wdees   
4                                 Die dritte Bibliothek        szylla   
...                                                 ...           ...   
1995  Staat, multinationale Konzerne und politische ...  fbw_hannover   
1996  Demokratische Wirtschaftspolitik gegen Marktma...  fbw_hannover   
1997                        Für den Frieden produzieren  fbw_hannover   
1998                         Lateinische Umgangssprache  filologanoga   
1999                        Texte zur Theorie des Films         antje   

                                       description                 date  \
0     Treffpunkt Bibliothek 24. - 31. Oktober 2010  2010-07-18 21:30:25   
1                                                   2020-09-22 15:08:13   
2                                                   2005-06-22 15:33:30   
3                                                   2016-05-09 12:23:32   
4                 Bibliothek von Mircea Cartarescu  2007-12-03 10:43:42   
...                                            ...                  ...   
1995                                                2009-08-21 09:46:50   
1996                                                2009-08-21 09:46:50   
1997                                                2009-08-21 09:46:50   
1998                                                2011-08-05 10:55:22   
1999                                                2011-05-10 09:22:37   

               changeDate  count  ... imtm-inventarnummer eventtitle venue  \
0     2010-07-18 21:30:25      1  ...                 NaN        NaN   NaN   


Im folgenden überschreiben wir df_bibsonomy_selection mit der erzeugten Ansicht: 

In [72]:
df_bibsonomy_selection = df_bibsonomy[["label", "author", "year", "type"]]

In [73]:
df_bibsonomy_selection

label              author  \
0                     Bibliothek: Treffpunkt Bibliothek                 NaN   
1                                            Bibliothek                 NaN   
2                                            Bibliothek                 NaN   
3                             Bibliothek - Fit4Research                 NaN   
4                                 Die dritte Bibliothek                 NaN   
...                                                 ...                 ...   
1995  Staat, multinationale Konzerne und politische ...  [Heinz-Jürgen Axt]   
1996  Demokratische Wirtschaftspolitik gegen Marktma...                [ ?]   
1997                        Für den Frieden produzieren                 NaN   
1998                         Lateinische Umgangssprache     [J. B. Hofmann]   
1999                        Texte zur Theorie des Films                 NaN   

      year         type  
0      NaN     Bookmark  
1      NaN     Bookmark  
2      NaN     Bookmark  
3      NaN     Bookmark  
4      NaN     Bookmark  
...    ...          ...  
1995  1978  Publication  
1996  1981  Publication  
1997  1981  Publication  
1998  1926  Publication  
1999  1998  Publication  

[2000 rows x 4 columns]

### Daten Filtern

Unser DataFrame enthält noch viele Zeilen, die wir herausfiltern möchten. .loc (locate) hilft, das DataFrame nach Kriterien zu Filtern.

In [74]:
# Daten auswählen, wo in Spalte "author" ein Wert eingetragen ist
df_bibsonomy_selection = df_bibsonomy_selection.loc[df_bibsonomy_selection["author"].isnull() == False]

In [75]:
df_bibsonomy_selection

label  \
1000  BIB Fundgrube Fachinformation Grüne Bibliothek...   
1001  Bibliothek 2.0 : neue Perspektiven und Einsatz...   
1002    Service vor Ort? Die Bibliothek geht zum Nutzer   
1003  Wissenschaft online. Elektronisches Publiziere...   
1004               Bibliothek 2.0 -- Wird alles anders?   
...                                                 ...   
1993                    Geschichte der Arbeitslosigkeit   
1994                              QuickStart PostScript   
1995  Staat, multinationale Konzerne und politische ...   
1996  Demokratische Wirtschaftspolitik gegen Marktma...   
1998                         Lateinische Umgangssprache   

                                               author  year         type  
1000     [ Berufsverband Information Bibliothek e.V.]  2021  Publication  
1001  [Michaela Selbach, Bibliothek der Fachhochsch.]  2007  Publication  
1002                                 [Manuela Schulz]  2010  Publication  
1003                                   [Beate Tröger]  2000  Publication  
1004               [Patrick Danowski, Lambert Heller]  2007  Publication  
...                                               ...   ...          ...  
1993                                    [Frank Niess]  1979  Publication  
1994              [Florian Däumling, Daniel Sillescu]  1990  Publication  
1995                               [Heinz-Jürgen Axt]  1978  Publication  
1996                                             [ ?]  1981  Publication  
1998                                  [J. B. Hofmann]  1926  Publication  

[901 rows x 4 columns]

In [76]:
# Daten auswählen, wo in Spalte "type" der Wert "Publication" ist
df_bibsonomy_selection.loc[df_bibsonomy_selection.type == "Publication"]

label  \
1000  BIB Fundgrube Fachinformation Grüne Bibliothek...   
1001  Bibliothek 2.0 : neue Perspektiven und Einsatz...   
1002    Service vor Ort? Die Bibliothek geht zum Nutzer   
1003  Wissenschaft online. Elektronisches Publiziere...   
1004               Bibliothek 2.0 -- Wird alles anders?   
...                                                 ...   
1993                    Geschichte der Arbeitslosigkeit   
1994                              QuickStart PostScript   
1995  Staat, multinationale Konzerne und politische ...   
1996  Demokratische Wirtschaftspolitik gegen Marktma...   
1998                         Lateinische Umgangssprache   

                                               author  year         type  
1000     [ Berufsverband Information Bibliothek e.V.]  2021  Publication  
1001  [Michaela Selbach, Bibliothek der Fachhochsch.]  2007  Publication  
1002                                 [Manuela Schulz]  2010  Publication  
1003                                   [Beate Tröger]  2000  Publication  
1004               [Patrick Danowski, Lambert Heller]  2007  Publication  
...                                               ...   ...          ...  
1993                                    [Frank Niess]  1979  Publication  
1994              [Florian Däumling, Daniel Sillescu]  1990  Publication  
1995                               [Heinz-Jürgen Axt]  1978  Publication  
1996                                             [ ?]  1981  Publication  
1998                                  [J. B. Hofmann]  1926  Publication  

[901 rows x 4 columns]

Über ".drop_duplicates" lassen sich doppelte Einträge entfernen. [Dokumentation](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.drop_duplicates.html)

```
keep = "first"
```
lässt uns jeweils die erste Zeile der Zeilen, die mehrfach vorkommen behalten.

In [77]:
# Entfernen von allen Duplikaten in der Spalte "label". 
df_bibsonomy_selection = df_bibsonomy_selection.drop_duplicates(subset = "label", keep = "first")

In [78]:
df_bibsonomy_selection

label  \
1000  BIB Fundgrube Fachinformation Grüne Bibliothek...   
1001  Bibliothek 2.0 : neue Perspektiven und Einsatz...   
1002    Service vor Ort? Die Bibliothek geht zum Nutzer   
1003  Wissenschaft online. Elektronisches Publiziere...   
1004               Bibliothek 2.0 -- Wird alles anders?   
...                                                 ...   
1993                    Geschichte der Arbeitslosigkeit   
1994                              QuickStart PostScript   
1995  Staat, multinationale Konzerne und politische ...   
1996  Demokratische Wirtschaftspolitik gegen Marktma...   
1998                         Lateinische Umgangssprache   

                                               author  year         type  
1000     [ Berufsverband Information Bibliothek e.V.]  2021  Publication  
1001  [Michaela Selbach, Bibliothek der Fachhochsch.]  2007  Publication  
1002                                 [Manuela Schulz]  2010  Publication  
1003                                   [Beate Tröger]  2000  Publication  
1004               [Patrick Danowski, Lambert Heller]  2007  Publication  
...                                               ...   ...          ...  
1993                                    [Frank Niess]  1979  Publication  
1994              [Florian Däumling, Daniel Sillescu]  1990  Publication  
1995                               [Heinz-Jürgen Axt]  1978  Publication  
1996                                             [ ?]  1981  Publication  
1998                                  [J. B. Hofmann]  1926  Publication  

[842 rows x 4 columns]

### Daten säubern

Im Folgenden sehen wir ein Beispiel dafür, wie wir mit Problemen umgehen können, die auf unsaubere Daten zurückzuführen sind: 

In [79]:
# Filtern nach einem bestimmten Wert in der Spalte "year"
df_bibsonomy_selection.loc[df_bibsonomy_selection["year"] > 2010]

TypeError: '>' not supported between instances of 'str' and 'int'

Problem: Der Datentyp ist nicht wie erwartet numerisch, sonder hier hier: String!

In [80]:
# .unique() gibt alle paarweise verschiedenen Werte (alle Werte) eines Series Objektes aus
df_bibsonomy_selection["year"].unique()

array(['2021', '2007', '2010', '2000', '2006', '2019', '2018', '2015',
       '2020', '2013', '2008', '2011', '2022', '2005', '2016', '1988',
       '1995', '2009', '1990', '2002', '1983', '1996', ' 2006 ', '1974',
       '2003', '2014', '2001', '2012', '2017', '1981', '1989', '1780',
       '1997', '1987', '1994', '1999', '1953', '1991', '1998', '1973',
       '1993', '1992', '1966', '1964', '1978', '2004', '1931', '1939',
       '1928', '1829', '1808', '1975', '1963', '1895', '1930', '1979',
       'n/a', '1984', '1926', '1922', '1935', '1986', '1933', '1961',
       '1937', '1927', '1783', '1914', 'March 2010', '1925', '1959',
       '1941', '2015 ', '1924', '1967', '1969', '1970', '1982', '1980',
       '1612', '1913', '1952', '1550', '1976', '1918', '1965', '1972',
       '1977', '1971', '1968', '19971998', '1957', '1929', '1985', '1910',
       '1958', '1955', '1954', '1960', '1947 [EA 1785]', '1943',
       '1997 (EA 1655)', '1956', '1909', '1923', '24.04.2005', '1950',
       '

Wir sehen, dass hier nicht nur Zahlen als String enthalten sind, sondern mit z.B. '1997 (EA 1655)' oder 'March 2010' auch ganze Strings enthalten sind.

Über .apply kann eine Funktion auf jeden Wert einer Spalte angewendet werden. "axis=1" beschreibt, dass Zeilenweise vorgegangen werden soll und "lambda row:", dass row hier jede iteration, in diesem Fall jede Zeile beschreibt.

Wir schreiben eine neue Spalte "year_correct", in der über row.year.isdigit() überprüft wird, ob es der eintrag in "year" Zahl ist bzw. in eine Ganzzahl umgewandelt werden kann.

In [81]:
df_bibsonomy_selection["year_correct"] = df_bibsonomy_selection.apply(lambda row: row.year.isdigit(), axis=1)

/var/folders/np/k1z3zp213l3cr57_g6n4bth80000gn/T/ipykernel_18245/2263534620.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_bibsonomy_selection["year_correct"] = df_bibsonomy_selection.apply(lambda row: row.year.isdigit(), axis=1)


Die Warnung kann an dieser stelle ignoriert werden.

über "axis=1" sagen wir hier, dass wir über Zeilen gehen müchten und mit "lambda row:" sagen wir, dass wir diese als "row" bezeichnen.

Jetzt schmeißen wir alle Zeilen aus dem Datensatz, deren "year" Eintrag keine Zahl ist:

In [83]:
df_bibsonomy_selection.loc[df_bibsonomy_selection.year_correct == False]

label  \
1084  Googles Herausforderung, für eine europäische ...   
1339  Die spezifischen Probleme einer kleinen Biblio...   
1429  Der Standort des elektronischen Leseplatzes un...   
1495  Library Life: Werkstätten kulturwissenschaftli...   
1922              Grundlegung zur Metaphysik der Sitten   
1937             Elemente der Philosophie I: Der Körper   
1962  ULB Bonn - AG Informationskompetenz - Schulung...   

                                                 author            year  \
1084                              [Jean-Noël Jeanneney]           2006    
1339                                   [Antonius Gusik]             n/a   
1429                               [Eric W. Steinhauer]      March 2010   
1495  [Friedolin Krentel, Katja Barthel, Sebastian B...           2015    
1922                                    [Immanuel Kant]  1947 [EA 1785]   
1937                                    [Thomas Hobbes]  1997 (EA 1655)   
1962   [ Arbeitsgemeinschaft Informationskompetenz NRW]      24.04.2005   

             type  year_correct  
1084  Publication         False  
1339  Publication         False  
1429  Publication         False  
1495  Publication         False  
1922  Publication         False  
1937  Publication         False  
1962  Publication         False

In [84]:
df_bibsonomy_selection = df_bibsonomy_selection.loc[df_bibsonomy_selection.year_correct == True]

In [85]:
df_bibsonomy_selection

label  \
1000  BIB Fundgrube Fachinformation Grüne Bibliothek...   
1001  Bibliothek 2.0 : neue Perspektiven und Einsatz...   
1002    Service vor Ort? Die Bibliothek geht zum Nutzer   
1003  Wissenschaft online. Elektronisches Publiziere...   
1004               Bibliothek 2.0 -- Wird alles anders?   
...                                                 ...   
1993                    Geschichte der Arbeitslosigkeit   
1994                              QuickStart PostScript   
1995  Staat, multinationale Konzerne und politische ...   
1996  Demokratische Wirtschaftspolitik gegen Marktma...   
1998                         Lateinische Umgangssprache   

                                               author  year         type  \
1000     [ Berufsverband Information Bibliothek e.V.]  2021  Publication   
1001  [Michaela Selbach, Bibliothek der Fachhochsch.]  2007  Publication   
1002                                 [Manuela Schulz]  2010  Publication   
1003                                   [Beate Tröger]  2000  Publication   
1004               [Patrick Danowski, Lambert Heller]  2007  Publication   
...                                               ...   ...          ...   
1993                                    [Frank Niess]  1979  Publication   
1994              [Florian Däumling, Daniel Sillescu]  1990  Publication   
1995                               [Heinz-Jürgen Axt]  1978  Publication   
1996                                             [ ?]  1981  Publication   
1998                                  [J. B. Hofmann]  1926  Publication   

      year_correct  
1000          True  
1001          True  
1002          True  
1003          True  
1004          True  
...            ...  
1993          True  
1994          True  
1995          True  
1996          True  
1998          True  

[835 rows x 5 columns]

Wir können jetzt auch die Zahl in in der Spalte "year" über "astype" in einen Integer, einen Ganzzahlenwertt umwandel und in "year_number" speichern.

In [86]:
df_bibsonomy_selection["year_number"] = df_bibsonomy_selection["year"].astype(int)

/var/folders/np/k1z3zp213l3cr57_g6n4bth80000gn/T/ipykernel_18245/974826408.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_bibsonomy_selection["year_number"] = df_bibsonomy_selection["year"].astype(int)


In [88]:
# die Neue Spalte besteht aus Integer Werten
df_bibsonomy_selection.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 835 entries, 1000 to 1998
Data columns (total 6 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   label         835 non-null    object
 1   author        835 non-null    object
 2   year          835 non-null    object
 3   type          835 non-null    object
 4   year_correct  835 non-null    bool  
 5   year_number   835 non-null    int64 
dtypes: bool(1), int64(1), object(4)
memory usage: 40.0+ KB


Jetzt können wir nach Publikationen Filtern, die nach 2010 erschienen sind:

In [89]:
df_bibsonomy_selection_recent = df_bibsonomy_selection.loc[df_bibsonomy_selection.year_number >=2010]

In [90]:
df_bibsonomy_selection_recent

label  \
1000  BIB Fundgrube Fachinformation Grüne Bibliothek...   
1002    Service vor Ort? Die Bibliothek geht zum Nutzer   
1011  Akademische Soziale Netzwerke und ihre Auswirk...   
1012              Netzwerk „Grüne Bibliothek“ gegründet   
1013  Und wo sind hier die Bücher? Bibliothek der Zu...   
...                                                 ...   
1832         allegro-C : eine Software im Epochenwandel   
1833  Pro smartphone cross-platform development - iP...   
1861  Der Beschäftigungszuschuss. Quantitative und q...   
1957                              Technik in der Antike   
1975  Internationaler VDI-Kongress Getriebe in Fahrz...   

                                                 author  year         type  \
1000       [ Berufsverband Information Bibliothek e.V.]  2021  Publication   
1002                                   [Manuela Schulz]  2010  Publication   
1011                                [Thomas Nachreiner]  2019  Publication   
1012                                 [Berthold Schmitt]  2018  Publication   
1013                                    [Simon Strauss]  2015  Publication   
...                                                 ...   ...          ...   
1832                               [Bernhard Eversberg]  2012  Publication   
1833       [Lee Lundrigan, Vidal Graupera, Sarah Allen]  2010  Publication   
1861                                    [Philipp Fuchs]  2013  Publication   
1957                                    [Brigitte Cech]  2010  Publication   
1975  [ diverse,  \VDI Society Product and Process D...  2012  Publication   

      year_correct  year_number  
1000          True         2021  
1002          True         2010  
1011          True         2019  
1012          True         2018  
1013          True         2015  
...            ...          ...  
1832          True         2012  
1833          True         2010  
1861          True         2013  
1957          True         2010  
1975          True         2012  

[211 rows x 6 columns]

BONUS:

Einfache Möglichkeit, alle Zeilen mit NaN Einträgen zu löschen:

In [91]:
df_bibsonomy_selection_no_nan = df_bibsonomy_selection.dropna()

In [92]:
df_bibsonomy_selection_no_nan

label  \
1000  BIB Fundgrube Fachinformation Grüne Bibliothek...   
1001  Bibliothek 2.0 : neue Perspektiven und Einsatz...   
1002    Service vor Ort? Die Bibliothek geht zum Nutzer   
1003  Wissenschaft online. Elektronisches Publiziere...   
1004               Bibliothek 2.0 -- Wird alles anders?   
...                                                 ...   
1993                    Geschichte der Arbeitslosigkeit   
1994                              QuickStart PostScript   
1995  Staat, multinationale Konzerne und politische ...   
1996  Demokratische Wirtschaftspolitik gegen Marktma...   
1998                         Lateinische Umgangssprache   

                                               author  year         type  \
1000     [ Berufsverband Information Bibliothek e.V.]  2021  Publication   
1001  [Michaela Selbach, Bibliothek der Fachhochsch.]  2007  Publication   
1002                                 [Manuela Schulz]  2010  Publication   
1003                                   [Beate Tröger]  2000  Publication   
1004               [Patrick Danowski, Lambert Heller]  2007  Publication   
...                                               ...   ...          ...   
1993                                    [Frank Niess]  1979  Publication   
1994              [Florian Däumling, Daniel Sillescu]  1990  Publication   
1995                               [Heinz-Jürgen Axt]  1978  Publication   
1996                                             [ ?]  1981  Publication   
1998                                  [J. B. Hofmann]  1926  Publication   

      year_correct  year_number  
1000          True         2021  
1001          True         2007  
1002          True         2010  
1003          True         2000  
1004          True         2007  
...            ...          ...  
1993          True         1979  
1994          True         1990  
1995          True         1978  
1996          True         1981  
1998          True         1926  

[835 rows x 6 columns]

### Daten Verändern und Hinzufügen
Neben der gerade kennengelernten Methode zum schreiben einer Spalte gibt es noch andere Möglichkeiten, wie Daten in ein DataFrame geschrieben werden können.

In [93]:
# Spalten umbenennen 
df_bibsonomy_selection = df_bibsonomy_selection.rename(columns={"label": "title", "author": "authors"})
df_bibsonomy_selection

title  \
1000  BIB Fundgrube Fachinformation Grüne Bibliothek...   
1001  Bibliothek 2.0 : neue Perspektiven und Einsatz...   
1002    Service vor Ort? Die Bibliothek geht zum Nutzer   
1003  Wissenschaft online. Elektronisches Publiziere...   
1004               Bibliothek 2.0 -- Wird alles anders?   
...                                                 ...   
1993                    Geschichte der Arbeitslosigkeit   
1994                              QuickStart PostScript   
1995  Staat, multinationale Konzerne und politische ...   
1996  Demokratische Wirtschaftspolitik gegen Marktma...   
1998                         Lateinische Umgangssprache   

                                              authors  year         type  \
1000     [ Berufsverband Information Bibliothek e.V.]  2021  Publication   
1001  [Michaela Selbach, Bibliothek der Fachhochsch.]  2007  Publication   
1002                                 [Manuela Schulz]  2010  Publication   
1003                                   [Beate Tröger]  2000  Publication   
1004               [Patrick Danowski, Lambert Heller]  2007  Publication   
...                                               ...   ...          ...   
1993                                    [Frank Niess]  1979  Publication   
1994              [Florian Däumling, Daniel Sillescu]  1990  Publication   
1995                               [Heinz-Jürgen Axt]  1978  Publication   
1996                                             [ ?]  1981  Publication   
1998                                  [J. B. Hofmann]  1926  Publication   

      year_correct  year_number  
1000          True         2021  
1001          True         2007  
1002          True         2010  
1003          True         2000  
1004          True         2007  
...            ...          ...  
1993          True         1979  
1994          True         1990  
1995          True         1978  
1996          True         1981  
1998          True         1926  

[835 rows x 6 columns]

In [94]:
# Schreiben einen Wertes in alle Zeilen einer Spalte
df_bibsonomy_selection["collected_by"] = "Fabian"
df_bibsonomy_selection

title  \
1000  BIB Fundgrube Fachinformation Grüne Bibliothek...   
1001  Bibliothek 2.0 : neue Perspektiven und Einsatz...   
1002    Service vor Ort? Die Bibliothek geht zum Nutzer   
1003  Wissenschaft online. Elektronisches Publiziere...   
1004               Bibliothek 2.0 -- Wird alles anders?   
...                                                 ...   
1993                    Geschichte der Arbeitslosigkeit   
1994                              QuickStart PostScript   
1995  Staat, multinationale Konzerne und politische ...   
1996  Demokratische Wirtschaftspolitik gegen Marktma...   
1998                         Lateinische Umgangssprache   

                                              authors  year         type  \
1000     [ Berufsverband Information Bibliothek e.V.]  2021  Publication   
1001  [Michaela Selbach, Bibliothek der Fachhochsch.]  2007  Publication   
1002                                 [Manuela Schulz]  2010  Publication   
1003                                   [Beate Tröger]  2000  Publication   
1004               [Patrick Danowski, Lambert Heller]  2007  Publication   
...                                               ...   ...          ...   
1993                                    [Frank Niess]  1979  Publication   
1994              [Florian Däumling, Daniel Sillescu]  1990  Publication   
1995                               [Heinz-Jürgen Axt]  1978  Publication   
1996                                             [ ?]  1981  Publication   
1998                                  [J. B. Hofmann]  1926  Publication   

      year_correct  year_number collected_by  
1000          True         2021       Fabian  
1001          True         2007       Fabian  
1002          True         2010       Fabian  
1003          True         2000       Fabian  
1004          True         2007       Fabian  
...            ...          ...          ...  
1993          True         1979       Fabian  
1994          True         1990       Fabian  
1995          True         1978       Fabian  
1996          True         1981       Fabian  
1998          True         1926       Fabian  

[835 rows x 7 columns]

In [95]:
# Spalten löschen
df_bibsonomy_selection = df_bibsonomy_selection.drop(columns=['year'])
df_bibsonomy_selection

title  \
1000  BIB Fundgrube Fachinformation Grüne Bibliothek...   
1001  Bibliothek 2.0 : neue Perspektiven und Einsatz...   
1002    Service vor Ort? Die Bibliothek geht zum Nutzer   
1003  Wissenschaft online. Elektronisches Publiziere...   
1004               Bibliothek 2.0 -- Wird alles anders?   
...                                                 ...   
1993                    Geschichte der Arbeitslosigkeit   
1994                              QuickStart PostScript   
1995  Staat, multinationale Konzerne und politische ...   
1996  Demokratische Wirtschaftspolitik gegen Marktma...   
1998                         Lateinische Umgangssprache   

                                              authors         type  \
1000     [ Berufsverband Information Bibliothek e.V.]  Publication   
1001  [Michaela Selbach, Bibliothek der Fachhochsch.]  Publication   
1002                                 [Manuela Schulz]  Publication   
1003                                   [Beate Tröger]  Publication   
1004               [Patrick Danowski, Lambert Heller]  Publication   
...                                               ...          ...   
1993                                    [Frank Niess]  Publication   
1994              [Florian Däumling, Daniel Sillescu]  Publication   
1995                               [Heinz-Jürgen Axt]  Publication   
1996                                             [ ?]  Publication   
1998                                  [J. B. Hofmann]  Publication   

      year_correct  year_number collected_by  
1000          True         2021       Fabian  
1001          True         2007       Fabian  
1002          True         2010       Fabian  
1003          True         2000       Fabian  
1004          True         2007       Fabian  
...            ...          ...          ...  
1993          True         1979       Fabian  
1994          True         1990       Fabian  
1995          True         1978       Fabian  
1996          True         1981       Fabian  
1998          True         1926       Fabian  

[835 rows x 6 columns]

Wir ermitteln mit apply, wie viele Autoren in "authors" eingetragen sind:

In [96]:
# Benutzen von .apply() um eine Funktion auf Teile der Daten anzuwenden (und hier, um das Ergebnis in einer neuen Spalte einzutragen)
df_bibsonomy_selection["number_authors"] = df_bibsonomy_selection.apply( lambda row: len(row.authors), axis=1)
df_bibsonomy_selection

title  \
1000  BIB Fundgrube Fachinformation Grüne Bibliothek...   
1001  Bibliothek 2.0 : neue Perspektiven und Einsatz...   
1002    Service vor Ort? Die Bibliothek geht zum Nutzer   
1003  Wissenschaft online. Elektronisches Publiziere...   
1004               Bibliothek 2.0 -- Wird alles anders?   
...                                                 ...   
1993                    Geschichte der Arbeitslosigkeit   
1994                              QuickStart PostScript   
1995  Staat, multinationale Konzerne und politische ...   
1996  Demokratische Wirtschaftspolitik gegen Marktma...   
1998                         Lateinische Umgangssprache   

                                              authors         type  \
1000     [ Berufsverband Information Bibliothek e.V.]  Publication   
1001  [Michaela Selbach, Bibliothek der Fachhochsch.]  Publication   
1002                                 [Manuela Schulz]  Publication   
1003                                   [Beate Tröger]  Publication   
1004               [Patrick Danowski, Lambert Heller]  Publication   
...                                               ...          ...   
1993                                    [Frank Niess]  Publication   
1994              [Florian Däumling, Daniel Sillescu]  Publication   
1995                               [Heinz-Jürgen Axt]  Publication   
1996                                             [ ?]  Publication   
1998                                  [J. B. Hofmann]  Publication   

      year_correct  year_number collected_by  number_authors  
1000          True         2021       Fabian               1  
1001          True         2007       Fabian               2  
1002          True         2010       Fabian               1  
1003          True         2000       Fabian               1  
1004          True         2007       Fabian               2  
...            ...          ...          ...             ...  
1993          True         1979       Fabian               1  
1994          True         1990       Fabian               2  
1995          True         1978       Fabian               1  
1996          True         1981       Fabian               1  
1998          True         1926       Fabian               1  

[835 rows x 7 columns]

### Daten Analysieren

In [97]:
# groupby() nutzen, um Zeilen zu gruppieren und Methoden darauf anzuwenden
# count zählt Anzahl der Zeilen pro gruppe
df_bibsonomy_selection_recent.groupby("year_number").count().label

year_number
2010        22
2011        12
2012        15
2013        29
2014        14
2015        23
2016        16
2017        10
2018         8
2019        18
2020        15
2021        17
2022        11
19971998     1
Name: label, dtype: int64

der Wert "19971998" wurde in eine Zahl umgewandelt, ist hier aber sicher falsch. Wir möchten diesen Eintrag jetzt entfernen. 

TIPP:
Man könnte in etwa so vorgehen, wie wir nach Veröffentlichungen nach 2010 gefiltert haben:
```
df_bibsonomy_selection_recent = df_bibsonomy_selection.loc[df_bibsonomy_selection.year_number >=2010]
```

### Daten Zusammenführen
Wir können mit Pandas auch relativ leicht mehrere Datensätze zusammenführen.

In [99]:
import json
# (Bibsonomy) JSON Daten einlesen
data1 = json.load(open("data/search_Cologne.json"))
data2 = json.load(open("data/search_Media_Bias.json"))

In [100]:
# Unsere Daten liegen wieder in "items"
data1.keys()

dict_keys(['types', 'properties', 'items'])

In [101]:
# Erstellen der DataFrames aus den Listen der Dictionaries, die die Dateneinträge der Bibsonomy Suche darstellen
df_search1 = pd.DataFrame(data1["items"])
df_search2 = pd.DataFrame(data2["items"])

In [102]:
df_search1

type                                                 id  \
0       Bookmark  https://www.bibsonomy.org/url/9a261c8872ecb14c...   
1       Bookmark  https://www.bibsonomy.org/url/a16672be36fd00b3...   
2       Bookmark  https://www.bibsonomy.org/url/b0b96f3140af77e8...   
3       Bookmark  https://www.bibsonomy.org/url/a16f6b2e28923bb4...   
4       Bookmark  https://www.bibsonomy.org/url/46c0d406f2483fe1...   
..           ...                                                ...   
195  Publication  https://www.bibsonomy.org/bibtex/20b19659687c1...   
196  Publication  https://www.bibsonomy.org/bibtex/2e2b970a28b61...   
197  Publication  https://www.bibsonomy.org/bibtex/2b27d82a4a7ba...   
198  Publication  https://www.bibsonomy.org/bibtex/255d71a5590ef...   
199  Publication  https://www.bibsonomy.org/bibtex/2507cbfcdc0af...   

                                                  tags  \
0                           [cologne, theater, cinema]   
1                    [de, jobs, cologne, selbstaendig]   
2                                 [coworking, cologne]   
3    [GErmany, Dom, Kirche, landmark, church, yovis...   
4                   [schmap, julianausserhofer, photo]   
..                                                 ...   
195                                         [oldsaxon]   
196                                             [dblp]   
197  [fragrances, beauty, cologne, cosmetics, perfu...   
198  [fragrances, fashion, beauty, cologne, cosmeti...   
199  [fragrances, fashion, beauty, cologne, cosmeti...   

                            intraHash  \
0    9a261c8872ecb14c92fbe783c8e13710   
1    a16672be36fd00b3dadb76f773711961   
2    b0b96f3140af77e8a421023136c35a65   
3    a16f6b2e28923bb455110f4945815381   
4    46c0d406f2483fe1cc89b8ef61c31798   
..                                ...   
195  0b19659687c14feede575239005667b3   
196  e2b970a28b61dda9ce5920e0da7f00f7   
197  b27d82a4a7bad66dae4a16e4272269a6   
198  55d71a5590ef68a8c46242690efd83df   
199  507cbfcdc0afac234b36408120456211   

                                                 label               user  \
0                                      Choices cologne             sorkin   
1                                      coLabor Cologne             fsteeg   
2                                    Coworking Cologne             fsteeg   
3    The Cologne Cathedral - More than 600 Years of...         lysander07   
4                        Cologne - Schmap Travel Guide  julianausserhofer   
..                                                 ...                ...   
195                          Altsächsisches Wörterbuch        searocraeft   
196  Proceedings of the 8th International Workshop ...               dblp   
197               Best Baby Perfumes For Your New Born        colognelove   
198  Our Latest Collection Of Best Smelling YSL Col...        colognelove   
199  7 Best Vanilla Perfumes That Will Make Your He...        colognelove   

                                           description                 date  \
0                   cologne Cinema and Theater program  2006-06-27 14:13:55   
1    coLabor is a coworking space to be set up some...  2007-03-28 00:22:19   
2    Coworking Cologne. Kleines Coworking Space in ...  2013-09-15 15:39:14   
3    On August 14, 1880, after more than 600 years ...  2013-08-14 19:43:10   
4    I was delighted to hear that two of my photos ...  2007-12-02 23:10:39   
..                                                 ...                  ...   
195                                                     2012-12-25 04:17:33   
196                                                     2020-02-12 00:00:00   
197                                                     2022-02-28 12:29:34   
198                                                     2022-02-28 10:36:38   
199                                                     2022-02-28 12:34:26   

              changeDate  count  ... unique-id  doi date-modified date-added  \
0    2006-06-27

In [103]:
df_search2

type                                                 id  \
0       Bookmark  https://www.bibsonomy.org/url/bf08e62b2ae96961...   
1       Bookmark  https://www.bibsonomy.org/url/252294047b101bdb...   
2       Bookmark  https://www.bibsonomy.org/url/27d2cf6d927e3c14...   
3       Bookmark  https://www.bibsonomy.org/url/24a281d4e8e33f81...   
4       Bookmark  https://www.bibsonomy.org/url/0734714d0d35f8a0...   
..           ...                                                ...   
101  Publication  https://www.bibsonomy.org/bibtex/257415091921f...   
102  Publication  https://www.bibsonomy.org/bibtex/271d33d933ed6...   
103  Publication  https://www.bibsonomy.org/bibtex/2a7189be7a33c...   
104  Publication  https://www.bibsonomy.org/bibtex/2b2f7171a0bb3...   
105  Publication  https://www.bibsonomy.org/bibtex/257585621f7ab...   

                                  tags                         intraHash  \
0           [politics, culture, media]  bf08e62b2ae96961b03e948b7b1cf010   
1           [news, bias, metric, home]  252294047b101bdb695790f5bf0f4142   
2             [web, bias, media, vote]  27d2cf6d927e3c1472656f41a38e9d80   
3                        [fact, check]  24a281d4e8e33f81b68912946cd84f8e   
4    [Diigo_-_Jem_Fleming's_Bookmarks]  0734714d0d35f8a0d70882d1e2215666   
..                                 ...                               ...   
101                             [dblp]  57415091921ff4f533289739183fb742   
102                             [dblp]  71d33d933ed6799821397aa8c94ae608   
103                             [dblp]  a7189be7a33c43a0027b1d2793e322ab   
104                             [dblp]  b2f7171a0bb346b7b8fab3c214fb2022   
105                             [dblp]  57585621f7ab645ec129ad215499488a   

                                                 label           user  \
0              MediaBite / A Shot at Bias in the media  georgeroberts   
1    Breaking News Headlines and Media Bias | Groun...        bshanks   
2      Skewz.com - Media Bias - All sides of the story        bshanks   
3    AllSides | Balanced news via media bias rating...       heluking   
4                                      Unfiltered News     jemfleming   
..                                                 ...            ...   
101  Tracking Bias in News Sources Using Social Med...           dblp   
102  Promoting users' intention to share online hea...           dblp   
103  Topic bias in the social media - The represent...           dblp   
104  Gender Bias in Sharenting: Both Men and Women ...           dblp   
105  That's Not the Way It Is: How User-Generated C...           dblp   

                                           description                 date  \
0                                                       2007-10-18 19:06:13   
1                                                       2022-04-28 20:20:48   
2                                                       2016-12-12 10:00:54   
3                                                       2020-04-30 16:29:51   
4    "The primary media outlets in every country ar...  2012-10-09 01:05:21   
..                                                 ...                  ...   
101                                                     2020-01-28 00:00:00   
102                                                     2020-11-07 00:00:00   
103                                                     2020-02-06 00:00:00   
104                                                     2018-08-13 00:00:00   
105                                                     2020-10-01 00:00:00   

              changeDate  count  ...      pages               isbn numpages  \
0    2013-09-12 08:18:16      1  ...        NaN                NaN      NaN   
1    2022-04-28 20:20:55      1  ...        NaN                NaN      NaN   
2    2020-06-15 21:39:35      1  ...        NaN                NaN      NaN   
3    2020-05-10 17:45:19      1  ...        NaN                NaN      NaN   
4    2012-10-09 01:23:45 

Es gibt zwar verschienene Möglichkeiten, DataFrames zusammenzuführen, in diesem Fall funktioniert .concat am besten, das eine Liste von Dictionaries als Argument nimmt:

In [104]:
# Mit .concat(Liste von DataFrames) lassen sich DataFrames miteinander kombinieren. 
df_both = pd.concat([df_search1, df_search2])

Das entstandene DataFrame hat alle Zeilen der beiden DataFrames:

In [105]:
df_both

type                                                 id  \
0       Bookmark  https://www.bibsonomy.org/url/9a261c8872ecb14c...   
1       Bookmark  https://www.bibsonomy.org/url/a16672be36fd00b3...   
2       Bookmark  https://www.bibsonomy.org/url/b0b96f3140af77e8...   
3       Bookmark  https://www.bibsonomy.org/url/a16f6b2e28923bb4...   
4       Bookmark  https://www.bibsonomy.org/url/46c0d406f2483fe1...   
..           ...                                                ...   
101  Publication  https://www.bibsonomy.org/bibtex/257415091921f...   
102  Publication  https://www.bibsonomy.org/bibtex/271d33d933ed6...   
103  Publication  https://www.bibsonomy.org/bibtex/2a7189be7a33c...   
104  Publication  https://www.bibsonomy.org/bibtex/2b2f7171a0bb3...   
105  Publication  https://www.bibsonomy.org/bibtex/257585621f7ab...   

                                                  tags  \
0                           [cologne, theater, cinema]   
1                    [de, jobs, cologne, selbstaendig]   
2                                 [coworking, cologne]   
3    [GErmany, Dom, Kirche, landmark, church, yovis...   
4                   [schmap, julianausserhofer, photo]   
..                                                 ...   
101                                             [dblp]   
102                                             [dblp]   
103                                             [dblp]   
104                                             [dblp]   
105                                             [dblp]   

                            intraHash  \
0    9a261c8872ecb14c92fbe783c8e13710   
1    a16672be36fd00b3dadb76f773711961   
2    b0b96f3140af77e8a421023136c35a65   
3    a16f6b2e28923bb455110f4945815381   
4    46c0d406f2483fe1cc89b8ef61c31798   
..                                ...   
101  57415091921ff4f533289739183fb742   
102  71d33d933ed6799821397aa8c94ae608   
103  a7189be7a33c43a0027b1d2793e322ab   
104  b2f7171a0bb346b7b8fab3c214fb2022   
105  57585621f7ab645ec129ad215499488a   

                                                 label               user  \
0                                      Choices cologne             sorkin   
1                                      coLabor Cologne             fsteeg   
2                                    Coworking Cologne             fsteeg   
3    The Cologne Cathedral - More than 600 Years of...         lysander07   
4                        Cologne - Schmap Travel Guide  julianausserhofer   
..                                                 ...                ...   
101  Tracking Bias in News Sources Using Social Med...               dblp   
102  Promoting users' intention to share online hea...               dblp   
103  Topic bias in the social media - The represent...               dblp   
104  Gender Bias in Sharenting: Both Men and Women ...               dblp   
105  That's Not the Way It Is: How User-Generated C...               dblp   

                                           description                 date  \
0                   cologne Cinema and Theater program  2006-06-27 14:13:55   
1    coLabor is a coworking space to be set up some...  2007-03-28 00:22:19   
2    Coworking Cologne. Kleines Coworking Space in ...  2013-09-15 15:39:14   
3    On August 14, 1880, after more than 600 years ...  2013-08-14 19:43:10   
4    I was delighted to hear that two of my photos ...  2007-12-02 23:10:39   
..                                                 ...                  ...   
101                                                     2020-01-28 00:00:00   
102                                                     2020-11-07 00:00:00   
103                                                     2020-02-06 00:00:00   
104                                                     2018-08-13 00:00:00   
105                                                     2020-10-01 00:00:00   

              changeDate  count  ... language citeulike-article-id priority  \
0    2006-06-27 

BONUS INFO: Der resultierende DataFrame hat mehr Spalten als die beiden DataFrames, die wir verbunden haben, weil es exklusive Spalten gibt, die nur in jeweils einem der beiden DataFrames erscheinen. Wir können das mit List Comprehension überprüfen:

In [ ]:
# Spalten, die in beiden DataFrames sind:
[a for a in df_search1.columns.to_list() if a in df_search2.columns.to_list()]

In [ ]:
# Spalten, die nur im ersten DataFrame sind:
[a for a in df_search1.columns.to_list() if a not in df_search2.columns.to_list()]

In [ ]:
# Spalten, die nur im zweiten DataFrame sind:
[a for a in df_search2.columns.to_list() if a not in df_search1.columns.to_list()]

## Daten Exportieren
es gibt eine Reihe von Möglichkeiten, Daten mit Pandas zu exportieren. Im folgenden zeigen wir, wie wir unseren Datensatz als CSV und JSON Datei exportieren können.

### Export als CSV
Zum Exportieren der Daten als als CSV Datei kann die .to_csv() Methode verwendet werden. Es muss ein Name für die Ausgabedatei angegeben werden. Um Probleme mit den Kommata beim Einlesen zu vermeiden, kann über sep= ein anderer Seperator angegeben werden, der nicht oder weniger Wahrscheinlich in den Daten vorkommt (oft bietet sich hier ein Semikolon oder ein anderes, noch seltener genutzes Sonderzeichen an). Pandas nutzt aber in der Regel alle Möglichkeiten, die Seperatoren in Daten zu "Escapen", sodass diese beim Einlesen nicht als Seperatoren Genutzt werden.

In [107]:
df_bibsonomy_selection.to_csv("data/output.csv", sep=";")

In [108]:
pd.read_csv("data/output.csv", sep=";")

Unnamed: 0                                              title  \
0          1000  BIB Fundgrube Fachinformation Grüne Bibliothek...   
1          1001  Bibliothek 2.0 : neue Perspektiven und Einsatz...   
2          1002    Service vor Ort? Die Bibliothek geht zum Nutzer   
3          1003  Wissenschaft online. Elektronisches Publiziere...   
4          1004               Bibliothek 2.0 -- Wird alles anders?   
..          ...                                                ...   
830        1993                    Geschichte der Arbeitslosigkeit   
831        1994                              QuickStart PostScript   
832        1995  Staat, multinationale Konzerne und politische ...   
833        1996  Demokratische Wirtschaftspolitik gegen Marktma...   
834        1998                         Lateinische Umgangssprache   

                                               authors         type  \
0       [' Berufsverband Information Bibliothek e.V.']  Publication   
1    ['Michaela Selbach', 'Bibliothek der Fachhochs...  Publication   
2                                   ['Manuela Schulz']  Publication   
3                                     ['Beate Tröger']  Publication   
4               ['Patrick Danowski', 'Lambert Heller']  Publication   
..                                                 ...          ...   
830                                    ['Frank Niess']  Publication   
831            ['Florian Däumling', 'Daniel Sillescu']  Publication   
832                               ['Heinz-Jürgen Axt']  Publication   
833                                             [' ?']  Publication   
834                                  ['J. B. Hofmann']  Publication   

     year_correct  year_number collected_by  number_authors  
0            True         2021       Fabian               1  
1            True         2007       Fabian               2  
2            True         2010       Fabian               1  
3            True         2000       Fabian               1  
4            True         2007       Fabian               2  
..            ...          ...          ...             ...  
830          True         1979       Fabian               1  
831          True         1990       Fabian               2  
832          True         1978       Fabian               1  
833          True         1981       Fabian               1  
834          True         1926       Fabian               1  

[835 rows x 8 columns]

In [109]:
df_bibsonomy_selection.to_csv("data/output.csv")

In [110]:
pd.read_csv("data/output.csv")

Unnamed: 0                                              title  \
0          1000  BIB Fundgrube Fachinformation Grüne Bibliothek...   
1          1001  Bibliothek 2.0 : neue Perspektiven und Einsatz...   
2          1002    Service vor Ort? Die Bibliothek geht zum Nutzer   
3          1003  Wissenschaft online. Elektronisches Publiziere...   
4          1004               Bibliothek 2.0 -- Wird alles anders?   
..          ...                                                ...   
830        1993                    Geschichte der Arbeitslosigkeit   
831        1994                              QuickStart PostScript   
832        1995  Staat, multinationale Konzerne und politische ...   
833        1996  Demokratische Wirtschaftspolitik gegen Marktma...   
834        1998                         Lateinische Umgangssprache   

                                               authors         type  \
0       [' Berufsverband Information Bibliothek e.V.']  Publication   
1    ['Michaela Selbach', 'Bibliothek der Fachhochs...  Publication   
2                                   ['Manuela Schulz']  Publication   
3                                     ['Beate Tröger']  Publication   
4               ['Patrick Danowski', 'Lambert Heller']  Publication   
..                                                 ...          ...   
830                                    ['Frank Niess']  Publication   
831            ['Florian Däumling', 'Daniel Sillescu']  Publication   
832                               ['Heinz-Jürgen Axt']  Publication   
833                                             [' ?']  Publication   
834                                  ['J. B. Hofmann']  Publication   

     year_correct  year_number collected_by  number_authors  
0            True         2021       Fabian               1  
1            True         2007       Fabian               2  
2            True         2010       Fabian               1  
3            True         2000       Fabian               1  
4            True         2007       Fabian               2  
..            ...          ...          ...             ...  
830          True         1979       Fabian               1  
831          True         1990       Fabian               2  
832          True         1978       Fabian               1  
833          True         1981       Fabian               1  
834          True         1926       Fabian               1  

[835 rows x 8 columns]

### Export als JSON
Zum Exportieren der Daten als als JSON file kann die .to_json() Methode verwendet werden. Es muss ein Name für die Ausgabedatei angegeben werden. Zusätzlich kann es sein, dass mit orient="records" eine eintragsorientierte ausgabe spezifiziert werden muss.   

In [111]:
df_bibsonomy_selection.to_json("data/output.json", orient="records")

In [112]:
# check, if it look correct
with open("data/output.json", "r") as f:
    j_load = json.load(f)
    print(json.dumps(j_load, indent=2))

[
  {
    "title": "BIB Fundgrube Fachinformation Gr\u00fcne Bibliothek. BibInfo",
    "authors": [
      " Berufsverband Information Bibliothek e.V."
    ],
    "type": "Publication",
    "year_correct": true,
    "year_number": 2021,
    "collected_by": "Fabian",
    "number_authors": 1
  },
  {
    "title": "Bibliothek 2.0 : neue Perspektiven und Einsatzm\u00f6glichkeiten f\u00fcr wissenschaftliche Bibliotheken",
    "authors": [
      "Michaela Selbach",
      "Bibliothek der Fachhochsch."
    ],
    "type": "Publication",
    "year_correct": true,
    "year_number": 2007,
    "collected_by": "Fabian",
    "number_authors": 2
  },
  {
    "title": "Service vor Ort? Die Bibliothek geht zum Nutzer",
    "authors": [
      "Manuela Schulz"
    ],
    "type": "Publication",
    "year_correct": true,
    "year_number": 2010,
    "collected_by": "Fabian",
    "number_authors": 1
  },
  {
    "title": "Wissenschaft online. Elektronisches Publizieren in Bibliothek und Hochschule",
    "aut

### Weitere Dateiformate
Pandas ist in der Lage, noch eine Vielzahl anderer Dateiformate zu erzeugen (und einzulesen). Dazu gehören: [Excel](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.to_excel.html#pandas.DataFrame.to_excel), [XML](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.to_xml.html) und viele weitere. Eine vollständige Übersicht findet sich [hier](https://pandas.pydata.org/docs/reference/io.html).

## Ressourcen zu Pandas
- [offizieller User Guide](https://pandas.pydata.org/docs/user_guide/index.html#user-guide)
- [offizielle API](https://pandas.pydata.org/docs/reference/index.html#api) mit detaillierten Informationen zu allen Mothoden
- [W3 Schools Pandas Tutorial](https://www.w3schools.com/python/pandas/default.asp) mit vielen Übungen und Quizes, sowie detaillierten Erläuterungen und Beispielen
- [Pandas Cheat Sheet](https://pandas.pydata.org/Pandas_Cheat_Sheet.pdf) – eine kompakte Übersicht vieler oft genutzter Funktionen
